shell/platform/android/io/flutter/embedding/android/FlutterImageView.java

```diff
@@ -242,52 +245,74 @@ public void closeImageReader() {
+      if (currentBitmap != null) {
+        canvas.drawBitmap(currentBitmap, 0, 0, null);
+        isBitmapAvailableForRendering = false;
+      }
     }
   }
 
-  private void closeCurrentImage() {
-    // Close and clear the current image if any.
-    if (currentImage != null) {
-      currentImage.close();
-      currentImage = null;
+  // This method is run on the |sBackgroundHandler|.
+  @Override
+  public void onImageAvailable(ImageReader reader) {
+    // 1. `acquireLatestImage` will discard any images in the queue up to the most recent
+    // one.
+    // 2. `acquireLatestImage()` may return null if no new image is available.
+    // 3. There's no guarantee that `onDraw()` is called after `invalidate()`.
+    // For example, the device may not produce new frames if it's in sleep mode
+    // or some special Android devices so the calls to `invalidate()` queued up
+    // until the device produces a new frame.
+    // 4. While the engine will also stop producing frames, there is a race condition.
+    try (final Image image = reader.acquireLatestImage()) {
+      if (image == null) {
+        return;
+      }
+
+      Bitmap bitmap = convertImageToBitmap(image);
+      if (bitmap != null) {
+        setCurrentBitmap(bitmap);
+        sUiHandler.postAtFrontOfQueue(() -> invalidate());
+      }
     }
   }
 
-  @TargetApi(29)
-  private void updateCurrentBitmap() {
+  @VisibleForTesting
+  /*package*/ Bitmap convertImageToBitmap(@NonNull Image image) {
+    Bitmap bitmap = null;
     if (android.os.Build.VERSION.SDK_INT >= 29) {
-      final HardwareBuffer buffer = currentImage.getHardwareBuffer();
-      currentBitmap = Bitmap.wrapHardwareBuffer(buffer, ColorSpace.get(ColorSpace.Named.SRGB));
+      final HardwareBuffer buffer = image.getHardwareBuffer();
+      bitmap = Bitmap.wrapHardwareBuffer(buffer, ColorSpace.get(ColorSpace.Named.SRGB));
       buffer.close();
     } else {
-      final Plane[] imagePlanes = currentImage.getPlanes();
+      final Plane[] imagePlanes = image.getPlanes();
       if (imagePlanes.length != 1) {
-        return;
+        return null;
       }
 
       final Plane imagePlane = imagePlanes[0];
       final int desiredWidth = imagePlane.getRowStride() / imagePlane.getPixelStride();
-      final int desiredHeight = currentImage.getHeight();
-
-      if (currentBitmap == null
-          || currentBitmap.getWidth() != desiredWidth
-          || currentBitmap.getHeight() != desiredHeight) {
-        currentBitmap =
-            Bitmap.createBitmap(
-                desiredWidth, desiredHeight, android.graphics.Bitmap.Config.ARGB_8888);
-      }
+      final int desiredHeight = image.getHeight();
+
+      bitmap =
+          Bitmap.createBitmap(
+              desiredWidth, desiredHeight, android.graphics.Bitmap.Config.ARGB_8888);
       ByteBuffer buffer = imagePlane.getBuffer();
       buffer.rewind();
-      currentBitmap.copyPixelsFromBuffer(buffer);
+      bitmap.copyPixelsFromBuffer(buffer);
```

Comment:
This is the only piece that would benefit things, but without a benchmark and/or tracing to show that this actually helps things in real applications we're just adding overhead for newer devices.

It might make sense to add some tracing around this call so we could look at existing benchmarks/applications and see how long it's really taking. I'm not 100% convinced that this will benefit too much from parallelization though, because Android devices tend to have slower memory access, and making more threads do more memory related work will probably be bad in some cases.